pkg/utils/revision/revision_utils.go

```diff
@@ -0,0 +1,368 @@
+// to a new LeaderWorkerSet using a strategic merge patch to replace the saved state of the new LeaderWorkerSet.
+func NewRevision(lws *leaderworkerset.LeaderWorkerSet, revision int64, templateHash string) (*appsv1.ControllerRevision, error) {
+	patch, err := GetPatch(lws)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewControllerRevision(lws,
+		controllerKind,
+		map[string]string{
+			leaderworkerset.TemplateRevisionHashKey: templateHash,
+			leaderworkerset.SetNameLabelKey:         lws.Name,
+		},
+		runtime.RawExtension{Raw: patch},
+		revision), nil
+}
+
+// ApplyRevision returns a new LeaderWorkerSet constructed by restoring the state in revision to set. If the returned error
+// is nil, the returned LeaderWorkerSet is valid.
+func ApplyRevision(lws *leaderworkerset.LeaderWorkerSet, revision *appsv1.ControllerRevision) (*leaderworkerset.LeaderWorkerSet, error) {
+	// clone := lws.DeepCopy()
+	str := &bytes.Buffer{}
+	err := unstructured.UnstructuredJSONScheme.Encode(lws, str)
+	if err != nil {
+		return nil, err
+	}
+	patched, err := strategicpatch.StrategicMergePatch(str.Bytes(), revision.Data.Raw, lws)
+	if err != nil {
+		return nil, err
+	}
+	restoredLws := &leaderworkerset.LeaderWorkerSet{}
+	if err = json.Unmarshal(patched, restoredLws); err != nil {
+		return nil, err
+	}
+	return restoredLws, nil
+}
+
+// nextRevision finds the next valid revision number based on revisions. If the length of revisions
+// is 0 this is 1. Otherwise, it is 1 greater than the largest revision's Revision. This method
+// assumes that revisions has been sorted by Revision.
+func NextRevision(revisions []*appsv1.ControllerRevision) int64 {
+	count := len(revisions)
+	if count <= 0 {
+		return 1
+	}
+
+	max := int64(1)
+	for _, revision := range revisions {
+		if max < revision.Revision {
+			max = revision.Revision
+		}
+	}
+	return max + 1
+}
+
+// TruncateHistory cleans up all other controller revisions except the currentRevision.
+// currentRevision is the one that matches the templateHash that is passed
+func TruncateHistory(ctx context.Context, k8sClient client.Client, lws *leaderworkerset.LeaderWorkerSet, templateHash string) error {
+	controllerHistory := History{Client: k8sClient, context: ctx}
+	selector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{MatchLabels: map[string]string{
+		leaderworkerset.SetNameLabelKey: lws.Name,
+	}})
+	if err != nil {
+		return err
+	}
+	revisions, err := controllerHistory.ListControllerRevisions(lws, selector)
+	if err != nil {
+		return err
+	}
+
+	for i, revision := range revisions {
+		if revision.Labels[leaderworkerset.TemplateRevisionHashKey] != templateHash {
+			if err := controllerHistory.DeleteControllerRevision(revisions[i]); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// Sha1Hash accepts an input string and returns the 40 character SHA1 hash digest of the input string.
```

Comment:
do we need this?

Reply:
Yes, we use this function to create the GroupKey https://github.com/kubernetes-sigs/lws/blob/main/pkg/webhooks/pod_webhook.go#L176

Reply:
ok, then we need to move it somewhere else, this is not related to revision management.